lib/type_check/internals/parser.ex

```diff
@@ -0,0 +1,268 @@
+        vars = vars |> Enum.map(&convert(&1, ctx))
+        # add new key-value pairs into existing vars
+        vars = Enum.zip(var_names, vars) |> Map.new()
+        vars = Map.merge(ctx.vars, vars)
+        convert(spec, %{ctx | vars: vars})
+
+      {:error, _} ->
+        ctx.default
+    end
+  end
+
+  defp convert(_, ctx), do: ctx.default
+
+  @spec convert_type(atom() | nil, list() | atom(), Context.t()) :: TypeCheck.Type.t()
+  # basic types
+  defp convert_type(:any, [], _), do: B.any()
+  defp convert_type(:atom, [], _), do: B.atom()
+  defp convert_type(:boolean, [], _), do: B.boolean()
+  defp convert_type(:pid, [], _), do: B.pid()
+  defp convert_type(:no_return, [], _), do: B.none()
+  defp convert_type(:none, [], _), do: B.none()
+
+  # unsupported by type_check yet
+  defp convert_type(:reference, [], ctx), do: ctx.default
+  defp convert_type(:port, [], ctx), do: ctx.default
+
+  # bitstrings
+  defp convert_type(:binary, [], _), do: B.binary()
+
+  defp convert_type(:binary, [{:integer, _, prefix}, {:integer, _, unit}], _) do
+    if unit == 0 do
+      B.sized_bitstring(prefix)
+    else
+      B.sized_bitstring(prefix, unit)
+    end
+  end
+
+  defp convert_type(:bitstring, [], _), do: B.bitstring()
+
+  # numbers
+  defp convert_type(:float, [], _), do: B.float()
+  defp convert_type(:integer, [], _), do: B.integer()
+  defp convert_type(:neg_integer, [], _), do: B.neg_integer()
+  defp convert_type(:non_neg_integer, [], _), do: B.non_neg_integer()
+  defp convert_type(:number, [], _), do: B.number()
+  defp convert_type(:pos_integer, [], _), do: B.pos_integer()
+
+  defp convert_type(:range, [{:integer, _, left}, {:integer, _, right}], _),
+    do: B.range(left, right)
+
+  # aliases
+  defp convert_type(:term, [], _), do: B.term()
+  defp convert_type(:arity, [], _), do: B.arity()
+  defp convert_type(:mfa, [], _), do: B.mfa()
+  defp convert_type(:module, [], _), do: B.module()
+  defp convert_type(:nonempty_binary, [], _), do: B.nonempty_binary()
+  defp convert_type(:nonempty_bitstring, [], _), do: B.nonempty_bitstring()
+  defp convert_type(:byte, [], _), do: B.byte()
+  defp convert_type(:char, [], _), do: B.char()
+  defp convert_type(:node, [], _), do: B.atom()
+  defp convert_type(:string, [], _), do: B.charlist()
+
+  # shothands for generics
+  defp convert_type(:fun, [], _), do: B.function()
+  defp convert_type(:function, [], _), do: B.function()
+
+  defp convert_type(:fun, [{:type, _, :any}, ret_type], ctx),
+    do: B.function(convert(ret_type, ctx))
+
+  defp convert_type(:list, [], _), do: B.list()
+  # empty list
+  defp convert_type(nil, [], _), do: B.list()
+  defp convert_type(:map, :any, _), do: B.map()
+  defp convert_type(:map, [], _), do: B.fixed_map([])
+  defp convert_type(:nonempty_list, [], _), do: B.nonempty_list()
+  # improper lists are cursed, restrict it to regular lists
+  defp convert_type(:nonempty_maybe_improper_list, [], _), do: B.nonempty_list()
```

Comment:
They are, but there is a good reason for them (c.f. "IO lists"). However, TypeCheck itself also does not properly support them yet. Defaulting to lists definitely is better than defaulting to any here.

However, we will need to revisit this in the future. Please remind me (with an issue?) once this PR is merged.